ghostscript: fix broken rpath on Darwin

siriobalmelli opened the pull request:
Dynamic library name on Darwin contains only 'maj.min' eg "9.53";
the build however used $version to set rpath;
this broke on 2029ca37 when $version went from "9.52" to "9.53.3".

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

cole-h commented:
@ofborg eval

veprbl commented on pkgs/misc/ghostscript/default.nix:
```diff
@@ -103,8 +103,10 @@ stdenv.mkDerivation rec {
+  dylib_version = with builtins; elemAt (match "([0-9]+\.[0-9]+).*" version) 0;
   preFixup = lib.optionalString stdenv.isDarwin ''
-    install_name_tool -change libgs.dylib.${version} $out/lib/libgs.dylib.${version} $out/bin/gs
+    install_name_tool -change libgs.dylib.${dylib_version} $out/lib/libgs.dylib.${dylib_version} $out/bin/gs
```
```suggestion
    install_name_tool -change libgs.dylib.$dylib_version $out/lib/libgs.dylib.$dylib_version $out/bin/gs
```

veprbl commented on pkgs/misc/ghostscript/default.nix:
```diff
@@ -103,8 +103,10 @@ stdenv.mkDerivation rec {
   '';
 
+  # version string may have 3 elements; dynamic library naming only contains the first 2
+  dylib_version = with builtins; elemAt (match "([0-9]+\.[0-9]+).*" version) 0;
```
```suggestion
  dylib_version = lib.majorMinor version;
```

siriobalmelli committed:
ghostscript: fix broken rpath on Darwin

Dynamic library name on Darwin contains only 'maj.min' eg "9.53";
the build however used $version to set rpath;
this broke on 2029ca37 when $version went from "9.52" to "9.53.3".

Add a call to 'gs' in installCheckPhase,
to break the build if dylib issues arise in the future.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>
Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

siriobalmelli commented:
@veprbl thank you for help cleaning this up. Re-rolled as a single commit and force-pushed.

veprbl commented on pkgs/misc/ghostscript/default.nix:
```diff
@@ -103,8 +103,11 @@ stdenv.mkDerivation rec {
   preFixup = lib.optionalString stdenv.isDarwin ''
-    install_name_tool -change libgs.dylib.${version} $out/lib/libgs.dylib.${version} $out/bin/gs
+    install_name_tool -change libgs.dylib.$dylib_version $out/lib/libgs.dylib.$dylib_version $out/bin/gs
+    $out/bin/gs --version
```
This line doesn't belong to preFixup. You could move it to `postCheck` or remove.

siriobalmelli replied:
I understand what you mean!

Unfortunately `postCheck` happens before `preFixup` ... my first attempt was to put it there but it was failing.

I considered not including it, but then there is a silent build failure in the future if linkage issues change. So, in light of that I figured this would be the "lesser evil".

The small consolation is that at least all of this Darwin linkage business is kept in one place, which makes it a bit neater I guess.

Any advice on how to make this cleaner while still maintaining a sanity check? If you feel it's best not to bother I can remove it I suppose.

veprbl replied:
Good point. Consider:
```nix
doInstallCheck = true;
installCheckPhase = ''
  runHook preInstallCheck

  "$out"/bin/gs --version

  runHook postInstallCheck
'';
```

siriobalmelli replied:
Thank you, I appreciate the help.

Updated and force-pushed.

veprbl reviewed (approved):
LGTM
